pkg/metadata/inventories/inventories.go

```diff
@@ -51,21 +52,28 @@ var (
 	// For testing purposes
 	timeNow   = time.Now
 	timeSince = time.Since
-)
 
-const (
-	// CloudProviderMetatadaName is the field name to use to set the cloud
-	// provider name in the agent metadata.
-	CloudProviderMetatadaName = "cloud_provider"
+	// KnownAgentMetadata lists all known agent-metadata keys.  Setting a key
+	// not on this list will result in a runtime panic.
+	KnownAgentMetadata = map[string]struct{}{
+		// cloud_provider is the name of the local cloud provider
+		"cloud_provider": {},
+		// hostname_source is the source of the hostname property, such as `gce` or `azure`.
+		"hostname_source": {},
+	}
 )
 
 // SetAgentMetadata updates the agent metadata value in the cache
 func SetAgentMetadata(name string, value interface{}) {
-	agentCacheMutex.Lock()
-	defer agentCacheMutex.Unlock()
+	agentMetadataMutex.Lock()
+	defer agentMetadataMutex.Unlock()
+
+	if _, found := KnownAgentMetadata[name]; !found {
+		panic(fmt.Sprintf("Agent metadata key %s not defined", name))
```

Comment:
I'm a bit curious why we would choose a panic over an error

Reply:
It's an implementation error, rather than a runtime error.  Much like a null pointer dereference.  I agree it'd be nicer to catch this at runtime, though.

Reply:
Did you consider replacing the key of `KnownAgentMetadata ` from `string` to a Golang int enum with iota?

Reply:
The idea of this package was for any part of the agent to be able to send arbitrary metadata (like it's the case for checks). Having a list on known field is nice but I'm not sure I like the panic when we receive a unknown one.

On the other hand we don't currently plan to have unknown field name for the agentMetadata